### PR TITLE
pubsub action case notification changes

### PR DESCRIPTION
**Pull request by creativeappdevamit**
# What and why?

# How to test?

# Trello
https://trello.com/c/0x04n5sk/527-s34-replace-rabbit-with-pub-sub-sample-to-case

### CODE REVIEW

**Review by insacuri — CHANGES_REQUESTED** [review]
A few small changes but looks good

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java [hunk 1]
```diff
@@ -1,58 +1,62 @@
...
-  private static final String ACTION_OUTBOUND_EXCHANGE = "action-outbound-exchange";
-  private static final String ACTION_ROUTING_KEY = "Action.CaseNotificationHandled.binding";
-
-  @ServiceActivator(
-      inputChannel = "caseNotificationTransformed",
-      adviceChain = "caseNotificationRetryAdvice")
-  public void acceptNotification(final CaseNotification caseNotification) throws CTPException {
+  private static ProjectSubscriptionName getPubSubSubscription(
+      String project, String subscriptionId) {
+    log.info(
```
[the remark this answers is not in the document]
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java]: Don't want to log my subscription Id.

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java [hunk 2]
```diff
@@ -1,58 +1,62 @@
...
-        log.error("Can't send message", e);
-      }
-    }
+  @EventListener(ApplicationReadyEvent.class)
+  public void caseNotificationSubscription() throws IOException {
+    // Instantiate an asynchronous message receiver.
+    MessageReceiver receiver =
+        (PubsubMessage message, AckReplyConsumer consumer) -> {
+          // Handle incoming message, then ack the received message.
+          log.info("Id: " + message.getMessageId());
```
[the remark this answers is not in the document]
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java]: its modified to debug level

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java [hunk 3]
```diff
@@ -1,58 +1,62 @@
...
+          log.info("Data: " + message.getData().toStringUtf8());
+          try {
+            CaseNotification caseNotificationReceiver =
+                objectMapper.readValue(message.getData().toStringUtf8(), CaseNotification.class);
+            caseNotificationService.acceptNotification(caseNotificationReceiver);
+            consumer.ack();
+          } catch (final IOException | CTPException e) {
+            log.error("Something went wrong");
```
[the remark this answers is not in the document]
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java]: it changed with new updates refactor

File: src/main/java/uk/gov/ons/ctp/response/action/service/PubSub.java [hunk 4]
```diff
@@ -28,4 +33,32 @@ public Publisher printfilePublisher() throws IOException {
     return publisherSupplier(
         appConfig.getGcp().getProject(), appConfig.getGcp().getPrintFileTopic());
   }
+
+  public ProjectSubscriptionName getActionCaseNotificationSubscriptionName() {
+    String project = appConfig.getGcp().getProject();
+    String subscriptionId = appConfig.getGcp().getCaseNotificationSubscription();
+    log.info(
```
[the remark this answers is not in the document]
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/PubSub.java:63]: as above

File: src/main/java/uk/gov/ons/ctp/response/action/config/ThreadPoolConfig.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+@Configuration
+public class ThreadPoolConfig implements AsyncConfigurer {
+
+  @Override
+  public Executor getAsyncExecutor() {
+    ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+    executor.setCorePoolSize(5);
+    executor.setMaxPoolSize(10);
+    executor.setThreadNamePrefix("SampleAsync-");
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/response/action/config/ThreadPoolConfig.java:18]: is this thread name correct?

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java [hunk 6]
```diff
@@ -1,58 +1,67 @@
...
+            log.with(e)
+                .error(
+                    "Something went wrong while processing message received from PubSub for case notification");
+          }
+        };
+    Subscriber subscriber = pubSub.getActionCaseNotificationSubscriber(receiver);
+    // Start the subscriber.
+    subscriber.startAsync().awaitRunning();
+    log.with(pubSub.getActionCaseNotificationSubscriptionName().toString())
+        .info("Listening for action case notification messages on port");
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:65]: on port ???

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java [hunk 7]
```diff
@@ -1,58 +1,67 @@
...
+            consumer.ack();
+          } catch (final IOException | CTPException e) {
+            log.with(e)
+                .error(
+                    "Something went wrong while processing message received from PubSub for case notification");
+          }
+        };
+    Subscriber subscriber = pubSub.getActionCaseNotificationSubscriber(receiver);
+    // Start the subscriber.
+    subscriber.startAsync().awaitRunning();
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:63]: does this spin up a new thread and then exit the method or does it block?

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java [hunk 8]
```diff
@@ -1,58 +1,67 @@
...
+          log.with(message.getData().toString()).debug("Receiving data from PubSub ");
+          try {
+            CaseNotification caseNotificationReceiver =
+                objectMapper.readValue(message.getData().toStringUtf8(), CaseNotification.class);
+            caseNotificationService.acceptNotification(caseNotificationReceiver);
+            consumer.ack();
+          } catch (final IOException | CTPException e) {
+            log.with(e)
+                .error(
+                    "Something went wrong while processing message received from PubSub for case notification");
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:58]: should it nack() the message here?

File: src/main/java/uk/gov/ons/ctp/response/action/service/PubSub.java [hunk 9]
```diff
@@ -1,31 +1,98 @@
...
+      log.info("Returning Subscriber");
+      ExecutorProvider executorProvider =
+          InstantiatingExecutorProvider.newBuilder().setExecutorThreadCount(4).build();
+      // `setParallelPullCount` determines how many StreamingPull streams the subscriber will open
+      // to receive message. It defaults to 1. `setExecutorProvider` configures an executor for the
+      // subscriber to process messages. Here, the subscriber is configured to open 2 streams for
+      // receiving messages, each stream creates a new executor with 4 threads to help process the
+      // message callbacks. In total 2x4=8 threads are used for message processing.
+      return Subscriber.newBuilder(getActionCaseNotificationSubscriptionName(), receiver)
+          .setParallelPullCount(2)
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/response/action/service/PubSub.java:90]: I wonder if we sshould make this a configurable value, i.e. expose it via the deployment and configure with helm?

**Review by warrenbailey — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/ons/ctp/response/action/config/ThreadPoolConfig.java (hunk 5 above)
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/config/ThreadPoolConfig.java:18]: copy paste error :-)

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java (hunk 6 above)
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:65]: Old habbits .. will change it to subscription

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java (hunk 7 above)
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:63]: not sure

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java (hunk 8 above)
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:58]: negative ack.? don't think so I am expecting IOException or CTP exception to happen by line number 53 and the ack is line 54

File: src/main/java/uk/gov/ons/ctp/response/action/service/PubSub.java (hunk 9 above)
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/PubSub.java:90]: I don't think we want to that the standard we want for subscriber to use. I don't think we can increase it anyways.

File: src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java (hunk 8 above)
Comment (creativeappdevamit) [reply src/main/java/uk/gov/ons/ctp/response/action/message/CaseNotificationReceiver.java:58]: Ya I need to do that

**Review by warrenbailey — APPROVED** [review]

**Review by insacuri — APPROVED** [review]

**Pull request merged**